File: CS3100-OS/fa19-cs3100-lecturenotes/Module3/threads/threads.java
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;


public class threads {

    public static class Maths implements Runnable {
        private String name;

        public Maths(String name) {
            this.name = "Maths #" + name;
        }

        public void run() {
            Thread.currentThread().setName(this.name);
            int a = 0;
            while (true)
                a++;
        }
    }

    public static class SlowIO implements Runnable {
        private static final int BUFSIZ = 8192;

        private String name;
        private String dev;


        public SlowIO(String dev) {
            this.name = "Slow I/O thread [" + dev + "]";
            this.dev = dev;
        }

        public void run() {
            Thread.currentThread().setName(this.name);
            System.out.format("\n\nWasting time with a cutting-edge I/O-Bound algorithm on the slow data source %s\n%d bytes at a time\n",
                    this.dev, SlowIO.BUFSIZ);

            byte[] buffer = new byte[SlowIO.BUFSIZ];
            while (true) {
                try {
                    FileInputStream slow = new FileInputStream(this.dev);
                    slow.read(buffer);
                    slow.close();

                    FileOutputStream devNull = new FileOutputStream("/dev/null");
                    devNull.write(buffer);
                    devNull.close();
                }
                catch (Exception e) {
                    System.err.println(e);
                }
            }
        }
    }


    public static void main(String[] args) {
        String IO_DEV = "/dev/sda";
        String device;

        if (args.length >= 1)
            device = args[0];
        else
            device = IO_DEV;

        try {
            // Create an array of 6 thread objects running the the 'maths' method
            Thread[] threads = new Thread[6];
            for (int i = 0; i < 6; ++i) {
                threads[i] = new Thread(new Maths(String.format("%d", i)));
                threads[i].start();
                System.out.format("Spawning maths thread #%d\n", i);
            }

            Thread tio = new Thread(new SlowIO(device));
            tio.start();
            System.out.format("Spawning Slow I/O thread on device %s\n", device);

            System.out.println("Press Ctrl-C to quit");

            for (int i = 0; i < 6; ++i) {
                threads[i].join();
                System.out.format("Joined maths thread #%d\n", i);
            }

            tio.join();
            System.out.println("Joined Slow I/O thread");
        }
        catch (Exception e) {
            System.err.println(e);
        }
    }
}

File: CS3100-OS/fa19-cs3100-lecturenotes/Module3/threads/threads.cpp
#include <iostream>
#include <pthread.h>
#include <cstring>
#include <cstdlib>
#include <cstdio>
#include <errno.h>

// A function which increments the variable 'a' forever
void* maths(__attribute__((unused)) void *threadid) {
    int a = 0;
    while (true)
        a++;
    return (void*)NULL;
}



// A function which copies 8k from our slow I/O device to /dev/null
void* slowio(__attribute__((unused)) void* vp_dev) {
    char* dev = (char*)vp_dev;

    std::cout << "\n\nWasting time with a cutting-edge I/O-Bound algorithm on the slow data source "
        << dev << ",\n" << BUFSIZ << " bytes at a time.\n";

    char buffer[BUFSIZ];

    while (true) {
        // open a slow device for read
        FILE *slow = fopen(dev, "r");
        if (slow) {
            // read BUFSIZ bytes from slow device
            fread(buffer, BUFSIZ, sizeof(char), slow);
            // close the slow device
            fclose(slow);

            // open /dev/null for write
            FILE *null = fopen("/dev/null", "w");
            if (null) {
                // write BUFSIZ to null
                fwrite(buffer, BUFSIZ, sizeof(char), null);
                // close null
                fclose(null);
            }
        }
        else {
            std::cerr << "\nFailed to open " << dev << "!!\n";
            return (void*)NULL;
        }
    }

    return (void*)NULL;
}



int main(int argc, char** argv) {
    // variable for referring to our "slow" I/O device
    char* IO_DEV = (char*)"/dev/sda";
    char *device;

    if (argc > 1 )
        device = argv[1];
    else
        device = IO_DEV;


    // Create an array of 6 thread objects running the the 'maths' function
    pthread_t tmaths[6];
    for (int i = 0; i < 6; ++i)
        if (pthread_create(&(tmaths[i]), NULL, maths, NULL)) {
            perror("pthread_create() failed on maths");
            exit(EXIT_FAILURE);
        }
        else {
            std::cout << "Spawned maths thread #" << i << std::endl;
        }


    // Create a thread object running the 'slowio' function
    pthread_t tio;
    if (pthread_create(&tio, NULL, slowio, (void*)device)) {
        perror("pthread_create() failed on slowio");
        exit(EXIT_FAILURE);
    }
    else {
        std::cout << "Spawned Slow I/O thread on device " << device << std::endl;
    }

    std::cout << "Press Ctrl-C to quit\n";

    for (int i = 0; i < 6; ++i) {
        pthread_join(tmaths[i], NULL);
        std::cout << "Joined maths thread #" << i << std::endl;
    }
    pthread_join(tio, NULL);
    std::cout << "Joined Slow I/O thread\n";

    return 0;
}
